tests/test_write_events.py

```diff
@@ -0,0 +1,194 @@
+                SiteName.DALLAS_FREE_PRESS,
+            ],
+        ],
+        columns=[
+            FieldSnowplow.DVCE_SCREENHEIGHT,
+            FieldSnowplow.USERAGENT,
+            FieldSnowplow.NETWORK_USERID,
+            FieldSnowplow.PAGE_REFERRER,
+            FieldSnowplow.EVENT_NAME,
+            FieldSnowplow.REFR_SOURCE,
+            FieldSnowplow.PAGE_URLPATH,
+            FieldSnowplow.DVCE_SCREENWIDTH,
+            FieldSnowplow.SEMISTRUCT_FORM_CHANGE,
+            FieldSnowplow.SEMISTRUCT_FORM_SUBMIT,
+            FieldSnowplow.PP_YOFFSET_MAX,
+            FieldSnowplow.DERIVED_TSTAMP,
+            FieldSnowplow.DOMAIN_SESSIONIDX,
+            FieldSnowplow.EVENT_ID,
+            FieldSnowplow.REFR_MEDIUM,
+            FieldSnowplow.SEMISTRUCT_FORM_FOCUS,
+            FieldSnowplow.DOMAIN_USERID,
+            FieldSnowplow.DOC_HEIGHT,
+            FieldNew.SITE_NAME,
+        ],
+    )
+
+    df = ConvertFieldTypes(
+        fields_int={FieldSnowplow.DOMAIN_SESSIONIDX},
+        fields_float={
+            FieldSnowplow.DOC_HEIGHT,
+            FieldSnowplow.DVCE_SCREENHEIGHT,
+            FieldSnowplow.DVCE_SCREENWIDTH,
+            FieldSnowplow.PP_YOFFSET_MAX,
+        },
+        fields_datetime={FieldSnowplow.DERIVED_TSTAMP},
+        fields_categorical={FieldSnowplow.EVENT_NAME, FieldSnowplow.REFR_MEDIUM, FieldSnowplow.REFR_SOURCE},
+    )(df)
+
+    return df
+
+
+@pytest.fixture(scope="module")
+def df_duplicate_key(df) -> pd.DataFrame:
+    df = df.copy()
+    # Change last event's ID to that of the first. Since both events have the same
+    # site nam, the last one will fail key-uniqueness check
+    df.loc[df.shape[0] - 1, FieldSnowplow.EVENT_ID] = df.loc[0, FieldSnowplow.EVENT_ID]
+    return df
+
+
+@pytest.fixture(scope="module")
+def db_name() -> str:
+    # Assuming the local DB name is "postgres"
+    return "postgres"
+
+
+@pytest.fixture(scope="module")
+def engine(db_name) -> Engine:
+    return create_engine(get_conn_string(db_name))
+
+
+@pytest.fixture(scope="module")
+def Session(engine) -> sessionmaker:
+    return sessionmaker(engine)
+
+
+# ---------- HELPERS ----------
+@contextmanager
+def create_and_drop_tables(engine: Engine) -> Generator[None, None, None]:
+    """
+    Context manager to safely create and drop tables before and after each test.
+    """
+    SQLModel.metadata.create_all(engine)
+    # The try-finally block ensures tables are still dropped in the event of an exception
+    # (see: https://realpython.com/python-with-statement/#opening-files-for-writing-second-version)
+    try:
+        yield
+    finally:
+        SQLModel.metadata.drop_all(engine)
+
+
+# ---------- TESTS ----------
+@pytest.mark.integration
+def test_write_events(df, engine, Session) -> None:
+    with create_and_drop_tables(engine):
+        # Write mock events to mock DB
+        write_events(df, Session)
+
+        # Assert all rows were written
+        with Session() as session, session.begin():
```

Comment:
🔧 remove call to `begin`
```suggestion
        with session_factory() as session:
```